### ISSUE TITLE
System.AccessViolationException

### ISSUE THREAD

**Issue by XudongXD8Lin** [k2-fsa/sherpa-onnx#3253]
var config = new OfflineSpeakerDiarizationConfig();
            config.Segmentation.Pyannote.Model = segmentationModelFile;
            config.Segmentation.NumThreads = 1;
            config.Segmentation.Debug = 0;
            config.Segmentation.Provider = "cpu";
            config.Embedding.Model = embeddingModelFile;
            config.Embedding.NumThreads = 1;
            config.Embedding.Debug = 0;
            config.Embedding.Provider = "cpu";
            _diarization = new OfflineSpeakerDiarization(config);

 _diarization.Process(processAudio);

“System.AccessViolationException”类型的未经处理的异常在 sherpa-onnx.dll 中发生 
Attempted to read or write protected memory. This is often an indication that other memory is corrupt.

**Comment by csukuangfj** [comment]
the information is too limited to be helpful. sorry, we cannot help you in this case.

**Comment by csukuangfj** [comment]
Please see 
https://docs.elephantrobotics.com/docs/ultraArm_sorting_kit_cn/14-IssueFAQ/14.0-how_to_ask.html
for how to ask.

**Comment by XudongXD8Lin** [comment]
In one of my C# WPF .NET8 projects, I utilized the Nuget package: org.k2fsa.sherpa.onnx
When I was using the OfflineSpeakerDiarization class to distinguish between different speakers, an exception occurred:
An unhandled exception of type 'System.AccessViolationException' occurred in sherpa-onnx.dll
Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Below is the code for initializing the OfflineSpeakerDiarization class:

var config = new OfflineSpeakerDiarizationConfig();
config.Segmentation.Pyannote.Model = segmentationModelFile;
config.Segmentation.NumThreads = 1;
config.Segmentation.Debug = 0;
config.Segmentation.Provider = "cpu";
config.Embedding.Model = embeddingModelFile;
config.Embedding.NumThreads = 1;
config.Embedding.Debug = 0;
config.Embedding.Provider = "cpu";
_diarization = new OfflineSpeakerDiarization(config);

When the following code was called at a certain time, the aforementioned exception occurred:


OfflineSpeakerDiarizationSegment[] segments;
……
segments = _diarization.Process(audioSamples);
Among them, float[] audioSamples represents a recorded audio clip with an unfixed length, but it is a fixed multiple of _diarization.SampleRate, and the possibility of float.NaN data is excluded.
I tried many methods, such as adding object locks, data alignment, excluding non-numbers, single-thread processing, creating an OfflineSpeakerDiarization object separately each time, and so on. However, exceptions still occur.

The segmentationModelFile is using sherpa-onnx-pyannote-segmentation-3-0
The embeddingModelFile is using 3dspeaker_speech_campplus_sv_zh_en_16k-common_advanced.onnx

**Comment by csukuangfj** [comment]
config.Segmentation.Debug = 1;
config.Embedding.Debug = 1;

and post all logs.

**Comment by XudongXD8Lin** [comment]
D:\a\sherpa-onnx\sherpa-onnx\sherpa-onnx\c-api\c-api.cc:GetOfflineSpeakerDiarizationConfig:2608 OfflineSpeakerDiarizationConfig(segmentation=OfflineSpeakerSegmentationModelConfig(pyannote=OfflineSpeakerSegmentationPyannoteModelConfig(model="D:\Users\a\source\repos\b\b\bin\Debug\net8.0-windows\Models\speaker\sherpa-onnx-pyannote-segmentation-3-0\model.onnx"), num_threads=1, debug=True, provider="cpu"), embedding=SpeakerEmbeddingExtractorConfig(model="D:\Users\a\source\repos\b\b\bin\Debug\net8.0-windows\Models\speaker\3dspeaker_speech_campplus_sv_zh_en_16k-common_advanced.onnx", num_threads=1, debug=True, provider="cpu"), clustering=FastClusteringConfig(num_clusters=-1, threshold=0.5), min_duration_on=0.3, min_duration_off=0.5)

D:\a\sherpa-onnx\sherpa-onnx\sherpa-onnx\csrc\offline-speaker-segmentation-pyannote-model.cc:Init:77 onnx.quant.pre_process=onnxruntime.quant
sample_rate=16000
num_speakers=3
num_classes=7
powerset_max_classes=2
receptive_field_size=991
window_size=160000
receptive_field_shift=270
model_type=pyannote-segmentation-3.0
version=1
model_author=pyannote
maintainer=k2-fsa
url_1=https://huggingface.co/pyannote/segmentation-3.0
url_2=https://huggingface.co/csukuangfj/pyannote-models/tree/main/segmentation-3.0
license=https://huggingface.co/pyannote/segmentation-3.0/blob/main/LICENSE


D:\a\sherpa-onnx\sherpa-onnx\sherpa-onnx\csrc\speaker-embedding-extractor-impl.cc:GetModelType:53 language=Chinese-English
framework=3d-speaker
feature_normalize_type=global-mean
comment=This model is from iic/speech_campplus_sv_zh_en_16k-common_advanced
url=https://www.modelscope.cn/models/iic/speech_campplus_sv_zh_en_16k-common_advanced/summary
sample_rate=16000
output_dim=192
normalize_samples=1

D:\a\sherpa-onnx\sherpa-onnx\sherpa-onnx\csrc\speaker-embedding-extractor-model.cc:Init:84 language=Chinese-English
framework=3d-speaker
feature_normalize_type=global-mean
comment=This model is from iic/speech_campplus_sv_zh_en_16k-common_advanced
url=https://www.modelscope.cn/models/iic/speech_campplus_sv_zh_en_16k-common_advanced/summary
sample_rate=16000
output_dim=192
normalize_samples=1

Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at SherpaOnnx.OfflineSpeakerDiarization.SherpaOnnxOfflineSpeakerDiarizationProcess(IntPtr, Single[], Int32)
--------------------------------
   at SherpaOnnx.OfflineSpeakerDiarization.Process(Single[])
   at b.Services.SpeakerDiarizationService.ProcessBatchWithProgress(Single[], System.Action`1<Single>)
   at b.Services.SpeakerDiarizationService.ProcessBatch(Single[])
   at b.Services.SenseVoiceStreamProcessor.TryProactiveSplit(Single[], System.Guid, System.TimeSpan)
   at b.Services.SenseVoiceStreamProcessor.ProcessAudioBuffer()
   at b.Services.SenseVoiceStreamProcessor.ProcessRecognitionCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.TimerQueueTimer.Fire(Boolean)
   at System.Threading.TimerQueue.FireNextTimers()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()

**Comment by 3rd-Musketeer** [comment]
## Same crash on macOS ARM64 (SIGBUS / exit code 138)

**Environment:**
- macOS 26.3, Apple Silicon (M-series), arm64
- sherpa-onnx 1.12.35 (Python, via pip)
- Models: `sherpa-onnx-pyannote-segmentation-3-0/model.onnx` + `3dspeaker_speech_campplus_sv_zh-cn_16k-common.onnx`

**Reproduction:**
- `OfflineSpeakerDiarization.process(audio)` crashes with SIGBUS (exit code 138) on specific audio segments
- The crash is **content-dependent** and **reproducible** on the same audio
- Audio ≤ 10s always works; certain segments at 12–15s reliably crash
- Same segment works fine if truncated to 10s — the crash is triggered when audio exceeds `window_size` (160000 samples / 10s) and enters the multi-chunk code path
- Tried `num_threads=1`, `int8` segmentation model, fresh object per call — none help

**Workaround:**
We bypassed `OfflineSpeakerDiarization` entirely and built an equivalent pipeline using only `SpeakerEmbeddingExtractor`:

```python
import sherpa_onnx
import numpy as np
from sklearn.cluster import SpectralClustering

emb_config = sherpa_onnx.SpeakerEmbeddingExtractorConfig(
    model="3dspeaker_speech_campplus_sv_zh-cn_16k-common.onnx",
    num_threads=4,
)
extractor = sherpa_onnx.SpeakerEmbeddingExtractor(emb_config)

# 1. Window the audio (1.5s windows, 0.75s step)
dur, step = 1.5, 0.75
chunks = []
st = 0.0
while st + dur < audio_len + step:
    ed = min(st + dur, audio_len)
    chunks.append((st, ed))
    st += step

# 2. Extract embeddings per window
embeddings = []
for s, e in chunks:
    clip = audio[int(s * sr):int(e * sr)]
    stream = extractor.create_stream()
    stream.accept_waveform(sr, clip)
    if extractor.is_ready(stream):
        embeddings.append(np.array(extractor.compute(stream)))

# 3. Cluster
embeddings = np.stack(embeddings)
sim = embeddings @ embeddings.T
affinity = (sim + 1) / 2
np.fill_diagonal(affinity, 1.0)
labels = SpectralClustering(n_clusters=2, affinity="precomputed").fit_predict(affinity)
```

This avoids the multi-chunk segmentation + embedding pipeline inside `OfflineSpeakerDiarization` and works reliably on all audio lengths. The `SpeakerEmbeddingExtractor` itself has no issues.

**Comment by kairoi-llc** [comment]
I believe I've found the root cause of this, and opened #3826 with a fix.

It's a **heap-use-after-free** in `OfflineSpeakerDiarizationPyannoteImpl::ComputeEmbeddings()`, not anything about model config or threading. The matrix of embeddings is over-allocated (one row per `(chunk, speaker)` pair) and then compacted with an *aliased* self-assignment:

```cpp
ans = ans(seq, Eigen::placeholders::all);
```

Shrinking the row count makes Eigen resize the destination first, and the resize frees the old buffer — while the right-hand side still points into it. The compaction branch only runs when at least one embedding was discarded as non-finite, which is why it's content-dependent and reproduces on the same audio every time.

@3rd-Musketeer — your observation is what confirmed it for me:

> Audio ≤ 10s always works; certain segments at 12–15s reliably crash … the crash is triggered when audio exceeds `window_size` (160000 samples / 10s) and enters the multi-chunk code path

That's exactly what this bug predicts: `Process()` short-circuits single-chunk audio through `HandleOneChunkSpecialCase()` and never reaches `ComputeEmbeddings()`, so only multi-chunk audio can reach the aliased shrink. It also explains the platform differences in this thread — glibc/macOS malloc often hands the same block straight back so the read silently succeeds, while allocators that quarantine freed chunks fault hard. The `System.AccessViolationException` in the original report is this read seen from a .NET P/Invoke frame; the SIGBUS is the same thing on macOS.

Verified with ASan: 4 of 7 pathological fixtures fault before the fix, 0 after, and 12 realistic clips produce byte-identical speaker turns before and after. Minimal repro is a single `+inf` sample in a >10s clip. Details, ASan trace and the validation matrix are in the PR.

Worth noting for anyone hitting this today: your `SpeakerEmbeddingExtractor` workaround sidesteps it precisely because it never enters that code path.

In the interest of disclosure: this diagnosis and patch were produced by an AI agent (Claude); a human approved submitting it but did not independently re-derive the analysis, so please review accordingly. Happy to share the fixtures, the standalone Eigen-only reproduction, or the raw logs.
